crates/wac-graph/src/graph.rs

```diff
@@ -0,0 +1,1884 @@
+
+        let name = name.into();
+        if self.exports.contains_key(&name) {
+            return Err(Error::ExternAlreadyExists {
+                kind: ExternKind::Export,
+                name,
+            });
+        }
+
+        let map_reader_err = |e: BinaryReaderError| {
+            let msg = e.to_string();
+            Error::InvalidExternName {
+                name: name.to_string(),
+                kind: ExternKind::Export,
+                source: anyhow::anyhow!(
+                    "{msg}",
+                    msg = msg.strip_suffix(" (at offset 0x0)").unwrap_or(&msg)
+                ),
+            }
+        };
+
+        // Ensure that the given export name is a valid extern name
+        match ComponentName::new(&name, 0).map_err(map_reader_err)?.kind() {
+            ComponentNameKind::Hash(_)
+            | ComponentNameKind::Url(_)
+            | ComponentNameKind::Dependency(_) => {
+                return Err(Error::InvalidExternName {
+                    name: name.to_string(),
+                    kind: ExternKind::Export,
+                    source: anyhow::anyhow!("export name cannot be a hash, url, or dependency"),
+                });
+            }
+            _ => {}
+        };
+
+        self.node_data_mut(id).export = Some(name.clone());
+
+        let prev = self.exports.insert(name, id);
+        assert!(prev.is_none());
+        Ok(())
+    }
+
+    /// Unmarks the given node from export.
+    ///
+    /// The node cannot be a _type definition node_ as type definitions are
+    /// always exported.
+    pub fn unexport_node(&mut self, id: NodeId) -> GraphResult<()> {
+        self.get_node(id).ok_or(Error::InvalidPackageId)?;
+
+        let data = self.node_data_mut(id);
+        if let NodeKind::Definition = data.kind {
+            return Err(Error::MustExportDefinition);
+        }
+
+        if let Some(name) = data.export.take() {
+            let removed = self.exports.swap_remove(&name);
+            assert!(removed.is_some());
+        }
+
+        Ok(())
+    }
+
+    /// Removes a node from the graph.
+    ///
+    /// All incoming and outgoing edges of the node are also removed.
+    ///
+    /// If the node has aliases, the aliased nodes are also removed.
+    ///
+    /// Returns `true` if the node was removed, otherwise returns `false`.
+    pub fn remove_node(&mut self, node: NodeId) -> bool {
+        if !self.graph.remove_node(node) {
+            return false;
+        }
+
+        self.free_node(node, false);
+        true
+    }
+
+    /// Adds an *instantiation argument edge* to the graph.
+    ///
+    /// The target node must be an instantiation node.
+    ///
+    /// The source node must be type-compatible with the argument of the target node.
+    ///
+    /// The argument must be a valid import on the target and not already have an
+    /// incoming edge from a different source node.
+    pub fn add_argument_edge(
```

Comment:
I renamed it to `connect_argument` and `disconnect_argument`; I'm not sure if that's better or worse, though.

Reply:
This might be an example of where I'm thinking about things less like a graph than I should be, but the high order bit for me is answering question "an argument to *what*?". Mentioning "connection" or "edge" isn't as necessary to me as the graphness of it all isn't at the forefront of my mind. That's why `instantiation_argument` makes more sense to me as it answers the question I was interested in. 

That all being said, I can imagine with more advanced use cases that having the graph like terminology at the forefront might make things more clear, so I'm not necessarily pushing one way or the other, but hopefully it's clear now why I prefer `instantiation_argument`.

Reply:
I generally like to have a verb in the method name, like `alias_instance_export` has "alias" as the verb.

Perhaps just `set_instantiation_argument`?

Reply:
I went with `set_instantiation_argument`/`unset_instantiation_argument`; with the rearranged parameters, I think it's an improvement over `connect_argument`/`disconnect_argument`.